realm/realm-library/src/main/java/io/realm/SetValueOperator.java

```diff
@@ -1181,6 +1185,116 @@ boolean retainAllInternal(Collection<?> c) {
+        super(baseRealm, osSet, valueClass);
+    }
+
+    @Override
+    boolean add(@Nullable Mixed value) {
+        if (value == null){
+            value = Mixed.nullValue();
+        } else if(value.getType() == MixedType.OBJECT) {
+            RealmModel realmModel = value.asRealmModel(RealmModel.class);
+            boolean copyObject = CollectionUtils.checkCanObjectBeCopied(baseRealm, realmModel, valueClass.getName(), SET_TYPE);
+            RealmObjectProxy proxy = (RealmObjectProxy) ((copyObject) ? CollectionUtils.copyToRealm(baseRealm, realmModel) : realmModel);
+            value = Mixed.valueOf(proxy);
+        }
+
+        return osSet.addMixed(value.getNativePtr());
+    }
+
+    @Override
+    boolean containsInternal(@Nullable Object o) {
+        Mixed value;
+        if (o == null) {
+            value = Mixed.nullValue();
+        } else {
+            value = (Mixed) o;
+        }
+        checkValidObject(value);
+        return osSet.containsMixed(value.getNativePtr());
+    }
+
+    @Override
+    boolean removeInternal(@Nullable Object o) {
+        // Object has been type-checked from caller
+        Mixed value;
+        if (o == null) {
+            value = Mixed.nullValue();
+        } else {
+            value = (Mixed) o;
+        }
+        checkValidObject(value);
+        return osSet.removeMixed(value.getNativePtr());
+    }
+
+    private void checkValidObject(Mixed mixed){
+        try {
+            mixed.checkValidObject(baseRealm);
+        } catch (IllegalArgumentException exception) {
+            throw new IllegalArgumentException("Mixed collection contains unmanaged objects.", exception);
+        }
+    }
+
+    @NotNull
+    private NativeMixedCollection getNativeMixedCollection(Collection<? extends Mixed> mixedCollection) {
+        long[] mixedPtrs = new long[mixedCollection.size()];
+        boolean[] notNull = new boolean[mixedCollection.size()];
+
+        int i = 0;
+        for (Mixed mixed : mixedCollection) {
+            if (mixed != null) {
+                checkValidObject(mixed);
+                mixedPtrs[i] = mixed.getNativePtr();
+                notNull[i] = true;
+            }
+            i++;
+        }
+
+        return NativeMixedCollection.newMixedCollection(mixedPtrs, notNull);
+    }
+
+    @Override
+    boolean containsAllInternal(Collection<?> c) {
+        // Collection has been type-checked from caller
+        //noinspection unchecked
+        NativeMixedCollection collection = getNativeMixedCollection((Collection<Mixed>) c);
+        return osSet.collectionFunnel(collection, OsSet.ExternalCollectionOperation.CONTAINS_ALL);
+    }
+
+    @Override
+    boolean addAllInternal(Collection<? extends Mixed> collection) {
+        // Collection has been type-checked from caller
+        // Use add method as it contains all the necessary checks
+        boolean result = false;
+        for (Mixed model : collection) {
```

Comment:
Is this OK performance wise compared to checking all of the objects and doing a single JNI roundtrip instead?

Reply:
Thats true, I believe we can do better